src/passes/SafeHeap.cpp

```diff
@@ -0,0 +1,316 @@
+    if (!alignfault.is()) {
+      auto* import = new Import;
+      import->name = alignfault = ALIGNFAULT_IMPORT;
+      import->module = ENV;
+      import->base = ALIGNFAULT_IMPORT;
+      import->kind = ExternalKind::Function;
+      import->functionType = ensureFunctionType("v", module)->name;
+      module->addImport(import);
+    }
+  }
+
+  void addGlobals(Module* module) {
+    // load funcs
+    Load load;
+    for (auto type : { i32, i64, f32, f64 }) {
+      load.type = type;
+      for (Index bytes : { 1, 2, 4, 8 }) {
+        load.bytes = bytes;
+        if (bytes > getWasmTypeSize(load.type)) continue;
+        for (auto signed_ : { true, false }) {
+          load.signed_ = signed_;
+          if (isWasmTypeFloat(load.type) && load.signed_) continue; 
+          for (Index align : { 1, 2, 4, 8 }) {
+            load.align = align;
+            if (align > bytes) continue;
+            for (auto isAtomic : { true, false }) {
+              load.isAtomic = isAtomic;
+              if (load.isAtomic && align != bytes) continue;
+              if (load.isAtomic && !module->memory.shared) continue;
+              addLoadFunc(&load, module);
+            }
+          }
+        }
+      }
+    }
+    // store funcs
+    Store store;
+    for (auto valueType : { i32, i64, f32, f64 }) {
+      store.valueType = valueType;
+      store.type = none;
+      for (Index bytes : { 1, 2, 4, 8 }) {
+        store.bytes = bytes;
+        if (bytes > getWasmTypeSize(store.valueType)) continue;
+        for (Index align : { 1, 2, 4, 8 }) {
+          store.align = align;
+          if (align > bytes) continue;
+          for (auto isAtomic : { true, false }) {
+            store.isAtomic = isAtomic;
+            if (store.isAtomic && align != bytes) continue;
+            if (store.isAtomic && !module->memory.shared) continue;
+            addStoreFunc(&store, module);
+          }
+        }
+      }
+    }
+  }
+
+  void addLoadFunc(Load* curr, Module* module) {
+    auto* func = new Function;
+    func->name = getLoadName(curr);
+    func->params.push_back(i32); // pointer
+    func->params.push_back(i32); // offset
+    func->vars.push_back(i32); // pointer + offset
+    func->result = curr->type;
+    Builder builder(*module);
+    auto* block = builder.makeBlock();
+    block->list.push_back(
+      builder.makeSetLocal(
+        2,
+        builder.makeBinary(
+          AddInt32,
+          builder.makeGetLocal(0, i32),
+          builder.makeGetLocal(1, i32)
+        )
+      )
+    );
+    // check for reading past valid memory: if pointer + offset + bytes
+    block->list.push_back(
+      makeBoundsCheck(curr->type, builder, 2)
+    );
+    // check proper alignment
+    if (curr->align > 1) {
+      block->list.push_back(
+        makeAlignCheck(curr->align, builder, 2)
+      );
+    }
+    // do the load
+    auto* load = module->allocator.alloc<Load>();
+    *load = *curr; // basically the same as the template we are given!
```

Comment:
Can we pull this out to the caller? It's non-obvious that the `curr` parameter is just a template, and by putting this copy in the same place as the callsite, it makes this usage less unusual.

Reply:
I'll add a comment to explain the Load/Store are used as structs to describe the type of load/store to create a function for. But I'm not sure I fully understand the question, what do you mean by "pull this out to the caller"?